src/line.rs

```diff
@@ -63,197 +62,154 @@ impl<'input, 'header, Endian> StateMachine<'input, 'header, Endian>
-            self.registers.op_index - (-operation_advance as u64)
-        } else {
-            self.registers.op_index + (operation_advance as u64)
-        };
+        let op_index_with_advance = self.row.op_index + operation_advance;
 
-        self.registers.address = self.registers.address +
-                                 minimum_instruction_length *
-                                 (op_index_with_advance / maximum_operations_per_instruction);
+        self.row.address = self.row.address +
+                           minimum_instruction_length *
+                           (op_index_with_advance / maximum_operations_per_instruction);
 
-        self.registers.op_index = op_index_with_advance % maximum_operations_per_instruction;
+        self.row.op_index = op_index_with_advance % maximum_operations_per_instruction;
     }
 
-    fn adjust_opcode(&self, opcode: u8) -> i64 {
-        let opcode = opcode as i64;
-        let line_base = self.header.line_base as i64;
-        opcode - line_base
+    fn adjust_opcode(&self, opcode: u8) -> u8 {
+        opcode - self.row.header.opcode_base
     }
 
     /// Section 6.2.5.1
-    fn exec_special_opcode(&mut self, opcode: u8) -> LineNumberRow<'input, 'header, Endian> {
+    fn exec_special_opcode(&mut self, opcode: u8) {
         let adjusted_opcode = self.adjust_opcode(opcode);
 
         // Step 1
 
-        let line_base = self.header.line_base as i64;
-        let line_range = self.header.line_range as i64;
-        let line_increment = line_base + (adjusted_opcode % line_range);
-        if line_increment < 0 {
-            let decrement = -line_increment as u64;
-            if decrement <= self.registers.line {
-                self.registers.line -= decrement;
-            } else {
-                self.registers.line = 0;
-            }
-        } else {
-            self.registers.line += line_increment as u64;
-        }
+        let line_base = self.row.header.line_base as i64;
+        let line_range = self.row.header.line_range;
+        let line_increment = line_base + (adjusted_opcode % line_range) as i64;
+        self.apply_line_advance(line_increment);
 
         // Step 2
 
-        let operation_advance = (adjusted_opcode as i64) / (self.header.line_range as i64);
-        self.apply_operation_advance(operation_advance);
-
-        // Step 3
-
-        let row = LineNumberRow::new(self.header, self.registers);
-
-        // Step 4
-
-        self.registers.basic_block = false;
-
-        // Step 5
-
-        self.registers.prologue_end = false;
-
-        // Step 6
-
-        self.registers.epilogue_begin = false;
-
-        // Step 7
-
-        self.registers.discriminator = 0;
-
-        row
+        let operation_advance = adjusted_opcode / self.row.header.line_range;
+        self.apply_operation_advance(operation_advance as u64);
     }
 
     /// Execute the given opcode, if a new row in the line number matrix is
     /// generated, return it.
```

Comment:
This comment needs to be updated. Also, not sure whether `Option<()>` is the best return type. Not sure if `bool` is any more satisfying. Maybe a custom `enum`? Can't really think of a good name for one...